src/core/server/saved_objects/export/point_in_time_finder.ts

```diff
@@ -0,0 +1,192 @@
+ * to call `close()` to clean up the PIT and prevent Elasticsearch from
+ * consuming resources unnecessarily. This will automatically be done for
+ * you if you reach the last page of results.
+ *
+ * @example
+ * ```ts
+ * const findOptions: SavedObjectsFindOptions = {
+ *   type: 'visualization',
+ *   search: 'foo*',
+ *   perPage: 100,
+ * };
+ *
+ * const finder = createPointInTimeFinder({
+ *   logger,
+ *   savedObjectsClient,
+ *   findOptions,
+ * });
+ *
+ * const responses: SavedObjectFindResponse[] = [];
+ * for await (const response of finder.find()) {
+ *   responses.push(...response);
+ *   if (doneSearching) {
+ *     await finder.close();
+ *   }
+ * }
+ * ```
+ */
+export function createPointInTimeFinder({
+  findOptions,
+  logger,
+  savedObjectsClient,
+}: {
+  findOptions: SavedObjectsFindOptions;
+  logger: Logger;
+  savedObjectsClient: SavedObjectsClientContract;
+}) {
+  return new PointInTimeFinder({ findOptions, logger, savedObjectsClient });
+}
+
+/**
+ * @internal
+ */
+export class PointInTimeFinder {
+  readonly #log: Logger;
+  readonly #savedObjectsClient: SavedObjectsClientContract;
+  readonly #findOptions: SavedObjectsFindOptions;
+  #open: boolean = false;
+  #pitId?: string;
+
+  constructor({
+    findOptions,
+    logger,
+    savedObjectsClient,
+  }: {
+    findOptions: SavedObjectsFindOptions;
+    logger: Logger;
+    savedObjectsClient: SavedObjectsClientContract;
+  }) {
+    this.#log = logger;
+    this.#savedObjectsClient = savedObjectsClient;
+    this.#findOptions = {
+      // Default to 1000 items per page as a tradeoff between
+      // speed and memory consumption.
+      perPage: 1000,
+      ...findOptions,
+    };
+  }
+
+  async *find() {
+    if (this.#open) {
+      throw new Error(
+        'Point In Time has already been opened for this finder instance. ' +
+          'Please call `close()` before calling `find()` again.'
+      );
+    }
+
+    // Open PIT and request our first page of hits
+    await this.open();
+
+    let lastResultsCount: number;
+    let lastHitSortValue: unknown[] | undefined;
+    do {
+      const results = await this.findNext({
+        findOptions: this.#findOptions,
+        id: this.#pitId,
+        ...(lastHitSortValue ? { searchAfter: lastHitSortValue } : {}),
+      });
+      this.#pitId = results.pit_id;
+      lastResultsCount = results.saved_objects.length;
+      lastHitSortValue = this.getLastHitSortValue(results);
+
+      this.#log.debug(`Collected [${lastResultsCount}] saved objects for export.`);
+
+      // Close PIT if this was our last page
+      if (this.#pitId && lastResultsCount < this.#findOptions.perPage!) {
+        await this.close();
+      }
+
+      yield results;
+      // We've reached the end when there are fewer hits than our perPage size,
+      // or when `close()` has been called.
+    } while (this.#open && lastHitSortValue && lastResultsCount >= this.#findOptions.perPage!);
```

Comment:
question: why do we need to check `lastHitSortValue` here? Can `res.saved_objects[res.saved_objects.length - 1].sort;` be undefined or is it only when `res.saved_objects.length < 1` that lastHitSortValue will be undefined?

If so it seems like `lastResultsCount >= this.#findOptions.perPage!` is already doing the same check?

Reply:
Ah, thanks for pointing this out. IIRC I needed to check that in the initial implementation, because otherwise `find` would be called without `searchAfter`, which would always return hits, resulting in an infinite loop.

But now that `this.#open` has been added I don't think it is necessarily any longer, so I went ahead and removed it.